<a href="https://colab.research.google.com/github/Dzaki020306/Pemrograman-Dasar/blob/main/Pertemuan%207.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
pimport numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:

Dzaki_Naufal_Ilham = pd.read_csv('/content/mahasiswa.csv')
#import file csv

In [ ]:
Dzaki_Naufal_Ilham.head()
#melihat 5 data teratas


nim   nama           jurusan    umur nilai_akhir  kehadiran_pct  \
0  E0001   Aldi            Teknik     150        59.5           82.0   
1  E0002   Budi            teknik    20.0        84.5           99.0   
2  E0003  Citra            Energi      -5        56.2           78.0   
3  E0004   Dewi            Teknik  twenty        63.0           99.0   
4  E0005    Eka  Manajemen Energi    22.0         999           65.0   

  tanggal_daftar  
0     2024-05-05  
1     2024-02-14  
2     2024-04-29  
3     2024-08-02  
4     2024-03-22

Fungsi head() digunakan untuk menampilkan beberapa baris pertama dari sebuah DataFrame, biasanya lima baris teratas. Tujuan utama dari penggunaan fungsi ini adalah untuk melihat struktur awal dataset dan memastikan bahwa data sudah terbaca dengan benar oleh program. Melalui hasil head(), kita dapat mengetahui apakah format penulisan data sudah konsisten, apakah ada kolom yang tidak sesuai, serta apakah terdapat nilai-nilai yang tidak wajar. Berdasarkan tampilan yang kamu berikan, bagian head() masih memiliki beberapa ketidaksesuaian seperti jurusan yang tidak konsisten dalam penggunaan huruf besar, nilai numerik yang berisi teks seperti “twenty”, serta terdapat nilai yang terlalu besar atau tidak logis seperti umur -5 dan nilai_akhir 999. Dengan demikian, head() sangat berguna sebagai langkah awal untuk mendeteksi kesalahan data sebelum dilakukan analisis lebih lanjut.

In [ ]:
Dzaki_Naufal_Ilham.tail()
#melihat 5 data terakhir

nim   nama      jurusan  umur nilai_akhir  kehadiran_pct tanggal_daftar
73  E0057  Fauzi       Energi  23.0        58.2           76.0     2024-02-05
74  E0039   Mita       Energi  23.0        77.2           79.0     2024-01-04
75  E0069  Salsa  Informatika  23.0        63.4           86.0     2024-04-14
76  E0031   Eros  Informatika  23.0        68.0           66.0     2024-02-27
77  E0026   Zara  Informatika  24.0        90.4           82.0     2024-11-24

Fungsi tail() digunakan untuk menampilkan beberapa baris terakhir dari DataFrame, sama seperti head() tetapi fokusnya pada bagian akhir dataset. Fungsi ini membantu pengguna memeriksa apakah data di bagian bawah tetap mengikuti format yang benar, apakah ada baris kosong, serta untuk memastikan tidak terjadi kesalahan input pada data terakhir. Dari hasil tail() yang kamu tampilkan, format data terlihat lebih rapi dan konsisten dibandingkan bagian awal. Nilai dalam kolom seperti umur, jurusan, dan nilai akhir sudah terlihat lebih logis dan format tanggal juga seragam. Dengan demikian, tail() sangat bermanfaat untuk memastikan kualitas data pada bagian akhir dataset, terutama ketika dataset memiliki jumlah baris yang besar.

In [ ]:
Dzaki_Naufal_Ilham.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78 entries, 0 to 77
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   nim             78 non-null     object 
 1   nama            78 non-null     object 
 2   jurusan         76 non-null     object 
 3   umur            76 non-null     object 
 4   nilai_akhir     76 non-null     object 
 5   kehadiran_pct   76 non-null     float64
 6   tanggal_daftar  76 non-null     object 
dtypes: float64(1), object(6)
memory usage: 4.4+ KB


Fungsi info() digunakan untuk menampilkan ringkasan struktur suatu DataFrame. Melalui hasil ini, kita dapat melihat jumlah baris data, jumlah kolom, tipe data pada setiap kolom, serta jumlah data yang tidak kosong atau non-null. Informasi ini sangat penting karena membantu kita memahami kualitas serta karakteristik dataset yang sedang digunakan. Berdasarkan tampilan output yang kamu berikan, dataset memiliki total 78 baris dan 7 kolom. Beberapa kolom seperti nim, nama, dan kehadiran_pct memiliki jumlah data lengkap, sementara kolom lainnya seperti jurusan, umur, nilai_akhir, dan tanggal_daftar memiliki data yang tidak sepenuhnya terisi. Selain itu, terlihat juga bahwa sebagian besar kolom masih bertipe object, yang menunjukkan bahwa data mungkin belum memiliki format numerik yang benar dan perlu dibersihkan atau dikonversi sebelum analisis lebih lanjut.

In [ ]:
Dzaki_Naufal_Ilham.isna().sum()

nim               0
nama              0
jurusan           2
umur              2
nilai_akhir       2
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Fungsi isna().sum() digunakan untuk menghitung jumlah nilai kosong (missing values) pada setiap kolom dalam DataFrame. Hasil dari fungsi ini memberikan gambaran yang jelas mengenai distribusi data yang hilang sehingga memudahkan dalam menentukan langkah penanganan seperti mengisi nilai yang hilang, menghapus baris tertentu, atau memperbaiki format data. Berdasarkan hasil yang ditampilkan, terdapat beberapa kolom yang memiliki nilai kosong yaitu kolom jurusan, umur, nilai_akhir, kehadiran_pct, dan tanggal_daftar, masing-masing sebanyak 2 data yang hilang. Sementara itu, kolom seperti nim dan nama tidak memiliki nilai kosong sama sekali. Informasi ini sangat penting agar proses pembersihan data dapat dilakukan secara lebih terarah dan tidak menimbulkan bias ketika analisis dilakukan.

In [ ]:

print("Jumlah duplikasi: ", Dzaki_Naufal_Ilham.duplicated().sum())

Jumlah duplikasi:  8


Perintah duplicated().sum() digunakan untuk menghitung jumlah baris data yang terdeteksi sebagai duplikat di dalam DataFrame. Duplikasi terjadi ketika terdapat dua atau lebih baris yang memiliki isi data yang sama persis pada seluruh kolom. Hasil dari pemeriksaan pertama menunjukkan bahwa terdapat 8 data yang terduplikasi. Informasi ini penting karena data ganda dapat mempengaruhi akurasi analisis dan menyebabkan bias dalam perhitungan statistik. Dengan mengetahui jumlah duplikasi, kita dapat memutuskan apakah data tersebut perlu dihapus atau diperiksa lebih lanjut sebelum melakukan analisis.

In [ ]:
Dzaki_Naufal_Ilham.drop_duplicates(inplace=True)

In [ ]:
print("Jumlah duplikasi: ", Dzaki_Naufal_Ilham.duplicated().sum())

Jumlah duplikasi:  0


Setelah mengetahui jumlah baris duplikat, fungsi drop_duplicates(inplace=True) digunakan untuk menghapus baris-baris duplikat tersebut langsung dari DataFrame. Penggunaan parameter inplace=True berarti perubahan dilakukan langsung pada dataset tanpa perlu membuat variabel baru. Setelah fungsi ini dijalankan, pemeriksaan ulang menggunakan duplicated().sum() menunjukkan bahwa jumlah duplikasi telah menjadi nol. Hal ini menunjukkan bahwa dataset sekarang sudah bersih dari data kembar, sehingga lebih siap untuk diproses dan dianalisis tanpa adanya pengulangan data yang tidak perlu.

In [ ]:
Dzaki_Naufal_Ilham.isna().sum()

nim               0
nama              0
jurusan           2
umur              2
nilai_akhir       2
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Langkah berikutnya adalah menjalankan kembali isna().sum() untuk mengecek apakah proses penghapusan duplikasi mempengaruhi jumlah data kosong pada dataset. Hasil yang ditampilkan menunjukkan bahwa kolom jurusan, umur, nilai_akhir, kehadiran_pct, dan tanggal_daftar masih memiliki masing-masing 2 nilai kosong, sementara kolom lainnya tidak memiliki missing value. Ini berarti bahwa penghapusan duplikasi tidak menambah jumlah data yang hilang, namun dataset tetap memerlukan penanganan lebih lanjut untuk membersihkan nilai kosong sebelum analisis dilakukan.

In [ ]:
Dzaki_Naufal_Ilham['jurusan'] = Dzaki_Naufal_Ilham['jurusan'].fillna("Pembangkit Energi")

In [ ]:
Dzaki_Naufal_Ilham.isna().sum()

nim               0
nama              0
jurusan           0
umur              2
nilai_akhir       2
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Langkah pertama yang dilakukan adalah memeriksa jumlah nilai kosong menggunakan fungsi isna().sum(). Hasil pemeriksaan tersebut menunjukkan bahwa terdapat beberapa kolom yang masih memiliki data kosong, yaitu nilai_akhir dengan 2 data hilang, kehadiran_pct dengan 2 data kosong, serta tanggal_daftar juga memiliki 2 nilai kosong. Sedangkan kolom nim, nama, jurusan, dan umur sudah tidak memiliki data kosong lagi setelah proses pembersihan sebelumnya. Tahapan ini penting untuk mengetahui bagian mana dari dataset yang perlu diperbaiki lebih lanjut sebelum dilakukan analisis atau pemrosesan lanjutan.

In [ ]:
Dzaki_Naufal_Ilham['umur'] = Dzaki_Naufal_Ilham['umur'].fillna(23)

Untuk memperbaiki masalah nilai kosong pada kolom umur, digunakan perintah fillna() dengan nilai 23 sebagai pengganti. Perintah tersebut dituliskan sebagai:Dzaki_Naufal_Ilham['umur'] = Dzaki_Naufal_Ilham['umur'].fillna(23)
Tujuan penggunaan fillna() adalah agar data yang sebelumnya kosong dapat terisi nilai baru sehingga kolom tersebut kembali lengkap tanpa nilai NaN. Pemilihan angka 23 kemungkinan didasarkan pada rata-rata, median, atau nilai yang dianggap paling representatif untuk kolom umur pada dataset. Dengan cara ini, dataset menjadi lebih stabil dan siap digunakan karena tidak ada data kosong yang dapat menyebabkan error pada perhitungan atau visualisasi berikutnya.

In [ ]:
Dzaki_Naufal_Ilham.isna().sum()

nim               0
nama              0
jurusan           0
umur              0
nilai_akhir       2
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Setelah proses pengisian nilai dilakukan, fungsi isna().sum() dijalankan kembali untuk memeriksa apakah nilai kosong pada kolom umur sudah berhasil diatasi. Hasil yang ditampilkan menunjukkan bahwa kolom umur kini memiliki nilai 0 missing value, artinya tidak ada lagi data kosong pada kolom tersebut. Sementara itu, nilai kosong pada kolom lainnya tetap sama karena proses pengisian baru diterapkan pada kolom umur saja. Tahapan ini memastikan bahwa perubahan yang dilakukan sudah berhasil dan dataset bergerak menuju kondisi yang lebih bersih dan lengkap.

In [ ]:
Dzaki_Naufal_Ilham['nilai_akhir'] = pd.to_numeric(Dzaki_Naufal_Ilham['nilai_akhir'], errors='coerce') #konversi tipe data menjadi numeric yaitu float
Dzaki_Naufal_Ilham['nilai_akhir'].mean() #mecari rata-rata

np.float64(88.15384615384616)

Penjelasan Konversi Tipe Data ke Numeric

Pada bagian ini, kolom nilai_akhir dikonversi menjadi tipe data numerik menggunakan fungsi pd.to_numeric(). Proses ini diperlukan karena sebelumnya kolom tersebut masih berisi campuran tipe data berupa angka dan teks, sehingga tidak dapat digunakan untuk perhitungan matematis atau statistik. Penggunaan parameter errors='coerce' membuat nilai yang tidak bisa dikonversi otomatis berubah menjadi NaN, sehingga lebih mudah terdeteksi sebagai data kosong yang perlu diperbaiki. Setelah konversi berhasil, proses selanjutnya adalah menghitung nilai rata-rata menggunakan .mean(), yang dalam hasil tersebut menghasilkan nilai sekitar 88.15, dan angka ini kemudian digunakan sebagai dasar pengisian data kosong selanjutnya.

In [ ]:
Dzaki_Naufal_Ilham.info()

<class 'pandas.core.frame.DataFrame'>
Index: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   nim             70 non-null     object 
 1   nama            70 non-null     object 
 2   jurusan         70 non-null     object 
 3   umur            70 non-null     object 
 4   nilai_akhir     65 non-null     float64
 5   kehadiran_pct   68 non-null     float64
 6   tanggal_daftar  68 non-null     object 
dtypes: float64(2), object(5)
memory usage: 4.4+ KB


Setelah melakukan konversi dan perhitungan mean, fungsi info() dijalankan kembali untuk memastikan bahwa tipe data pada kolom nilai_akhir telah berubah menjadi float64. Hal ini menunjukkan bahwa kolom tersebut sekarang sudah siap untuk dianalisis lebih lanjut secara numerik. Pada hasil info() juga terlihat adanya penurunan jumlah data non-null pada kolom nilai_akhir menjadi 68, yang berarti ada nilai yang sebelumnya berupa teks tidak valid dan telah berubah menjadi NaN. Tahapan verifikasi ini penting agar pengguna mengetahui perubahan struktur dataset yang terjadi setelah proses pembersihan dilakukan.

In [ ]:
Dzaki_Naufal_Ilham['nilai_akhir'] = Dzaki_Naufal_Ilham['nilai_akhir'].fillna(88)

In [ ]:
Dzaki_Naufal_Ilham.isna().sum()

nim               0
nama              0
jurusan           0
umur              0
nilai_akhir       0
kehadiran_pct     2
tanggal_daftar    2
dtype: int64

Tahapan berikutnya adalah mengisi data kosong pada kolom nilai_akhir menggunakan metode .fillna(), dengan nilai mean yang telah dihitung sebelumnya yaitu 88 (dibulatkan). Proses ini bertujuan untuk memastikan bahwa seluruh baris data memiliki nilai lengkap tanpa NaN, sehingga tidak terjadi error saat dilakukan analisis lebih lanjut seperti visualisasi, pengelompokan data, atau perhitungan statistik tambahan. Setelah nilai kosong terisi, pengecekan ulang menggunakan isna().sum() menunjukkan bahwa kolom nilai_akhir sudah tidak memiliki missing value lagi, sehingga proses pembersihan untuk kolom tersebut dinyatakan berhasil.

In [ ]:
Dzaki_Naufal_Ilham['kehadiran_pct'] = pd.to_numeric(Dzaki_Naufal_Ilham['kehadiran_pct'], errors='coerce') #konversi tipe data menjadi numeric yaitu float

Dzaki_Naufal_Ilham['kehadiran_pct'].mean() #mecari rata-rata

np.float64(82.13235294117646)

Tujuan: Langkah ini bertujuan untuk memastikan kolom 'kehadiran_pct' memiliki tipe data numerik.Proses: pd.to_numeric() digunakan untuk mengubah kolom. Opsi errors='coerce' akan mengganti semua nilai yang tidak bisa diubah menjadi angka dengan nilai $\text{NaN}$ (Not a Number), yang merupakan representasi nilai hilang.Output: Setelah konversi, rata-rata (mean) dari kolom tersebut dihitung:

In [ ]:
Dzaki_Naufal_Ilham['kehadiran_pct'] = Dzaki_Naufal_Ilham['kehadiran_pct'].fillna(82)
#sekarang isi missing value pada nilai_akhir dengan nilai rata-rata yg sudah di dapat sebelumnya

In [ ]:
Dzaki_Naufal_Ilham.isna().sum()
#cek hasilnya

nim               0
nama              0
jurusan           0
umur              0
nilai_akhir       0
kehadiran_pct     0
tanggal_daftar    2
dtype: int64

Tujuan: Memeriksa kembali berapa banyak nilai hilang yang tersisa di setiap kolom setelah melakukan imputasi pada kolom 'kehadiran_pct'.Hasil:| Kolom | Jumlah $\text{NaN}$ || :--- | :--- || nim | 0 || nama | 0 || jurusan | 0 || umur | 0 || nilai_akhir | 0 || kehadiran_pct | 0 || tanggal_daftar | 2 |Semua kolom selain 'tanggal_daftar' kini memiliki 0 nilai hilang. Kolom 'tanggal_daftar' masih memiliki 2 nilai hilang. Tipe data kolom ini adalah $int64$, tetapi isinya berupa tanggal dalam format string, yang terlihat dari langkah terakhir.

In [ ]:
Dzaki_Naufal_Ilham['tanggal_daftar'] = Dzaki_Naufal_Ilham['tanggal_daftar'].fillna('2024-03-24')

# karena tanggal lagir di sini bertipe string jadi langsung masukan saja tidak perlu di cari rata-rata nya atau konversi ke numerik


In [ ]:
Dzaki_Naufal_Ilham.isna().sum()

nim               0
nama              0
jurusan           0
umur              0
nilai_akhir       0
kehadiran_pct     0
tanggal_daftar    2
dtype: int64

Langkah kedua adalah pemeriksaan akhir untuk memverifikasi bahwa semua nilai yang hilang sudah tertangani:Kode ini menjalankan fungsi .isna().sum() pada DataFrame, yang menghitung total nilai $\text{NaN}$ yang tersisa di setiap kolom. Hasilnya ditampilkan dalam format Series sebagai berikut:
Tentu, saya akan menjelaskan dua bagian kode dan hasilnya yang terlihat pada gambar tersebut. Bagian ini melanjutkan proses penanganan nilai yang hilang (missing values) pada DataFrame.

📝 Penjelasan Langkah Penanganan Data
1. Imputasi Nilai Hilang pada Kolom Tanggal
Langkah pertama menangani sisa nilai yang hilang pada kolom 'tanggal_daftar':

Python
Dzaki_Naufal_Ilham['tanggal_daftar'] = Dzaki_Naufal_Ilham['tanggal_daftar'].fillna('2024-03-24')
# karena tanggal lahir di sini bertipe string jadi langsung masukan saja tidak perlu di cari rata-rata nya atau konversi ke numerik
Kode ini menggunakan metode .fillna('2024-03-24') untuk mengganti setiap nilai NaN (Not a Number) atau nilai kosong yang tersisa dalam kolom 'tanggal_daftar' dengan tanggal spesifik '2024-03-24'. Proses ini disebut imputasi hard-coding atau imputasi constant value. Dalam konteks pembersihan data, kolom yang berisi data non-numerik seperti tanggal, teks, atau kategori tidak dapat diisi menggunakan rata-rata, sehingga nilai pengganti harus ditetapkan secara manual berdasarkan kebijakan atau konteks data. Catatan di bawah kode memperjelas bahwa karena kolom ini bertipe string (teks yang mewakili tanggal), kita tidak perlu mencari rata-rata atau mengkonversinya ke numerik.

2. Pengecekan Ulang Hasil Imputasi
Langkah kedua adalah pemeriksaan akhir untuk memverifikasi bahwa semua nilai yang hilang sudah tertangani:

Python
Dzaki_Naufal_Ilham.isna().sum()
Kode ini menjalankan fungsi .isna().sum() pada DataFrame, yang menghitung total nilai NaN yang tersisa di setiap kolom. Hasilnya ditampilkan dalam format Series sebagai berikut:

Kolom	Jumlah NaN
nim	0
nama	0
jurusan	0
umur	0
nilai_akhir	0
kehadiran_pct	0
tanggal_daftar	2
Hasil pengecekan menunjukkan bahwa meskipun kolom 'tanggal_daftar' telah diisi dengan nilai konstan pada langkah sebelumnya, masih ada 2 nilai hilang yang terdeteksi. Ini bisa mengindikasikan bahwa nilai kosong di kolom tersebut mungkin tidak dalam format NaN yang dikenali oleh Pandas (seperti string kosong atau nilai spasi), atau proses fillna perlu dijalankan lagi. Meskipun demikian, semua kolom numerik dan kolom teks lain telah bersih dengan nilai 0 yang hilang, mengindikasikan sebagian besar proses pembersihan data sudah berhasil.

In [ ]:
Dzaki_Naufal_Ilham.jurusan.value_counts()

jurusan
Energi               21
Informatika          17
Teknik               12
Manajemen Energi     11
Pembangkit Energi     2
teknik                1
 energi               1
ManajemenEnergi       1
EnErGi                1
Teknik                1
ENG                   1
ENERGI                1
Name: count, dtype: int64


Tentu, saya akan menjelaskan maksud dari kode dan hasil yang ditampilkan pada gambar tersebut.

📊 Analisis Distribusi Nilai Kolom 'jurusan'
Kode yang dijalankan adalah Dzaki_Naufal_Ilham.jurusan.value_counts().

Tujuan Kode
Kode ini bertujuan untuk menghitung frekuensi atau jumlah kemunculan setiap nilai unik (value counts) dalam kolom 'jurusan' pada DataFrame yang bernama Dzaki_Naufal_Ilham. Ini adalah langkah penting dalam tahap eksplorasi data (Exploratory Data Analysis/EDA) dan pembersihan data (data cleansing), khususnya untuk mengidentifikasi masalah inkonsistensi penulisan data kategorikal.Hasilnya adalah tabel yang menunjukkan nilai unik dalam kolom 'jurusan' beserta jumlah kemunculannya (count):
Tentu, saya akan menjelaskan maksud dari kode dan hasil yang ditampilkan pada gambar tersebut.

📊 Analisis Distribusi Nilai Kolom 'jurusan'
Kode yang dijalankan adalah Dzaki_Naufal_Ilham.jurusan.value_counts().

Tujuan Kode
Kode ini bertujuan untuk menghitung frekuensi atau jumlah kemunculan setiap nilai unik (value counts) dalam kolom 'jurusan' pada DataFrame yang bernama Dzaki_Naufal_Ilham. Ini adalah langkah penting dalam tahap eksplorasi data (Exploratory Data Analysis/EDA) dan pembersihan data (data cleansing), khususnya untuk mengidentifikasi masalah inkonsistensi penulisan data kategorikal.

Hasil dan Interpretasi
Hasilnya adalah tabel yang menunjukkan nilai unik dalam kolom 'jurusan' beserta jumlah kemunculannya (count):

jurusan	count
Energi	21
Informatika	17
Teknik	12
Manajemen Energi	11
Pembangkit Energi	2
teknik	1
energi	1
ManajemenEnergi	1
EnErGi	1
Teknik	1
ENG	1
ENERGI	1
Masalah Inkonsistensi Data (Data Inconsistency)
Hasil perhitungan ini dengan jelas menunjukkan adanya inkonsistensi data yang serius pada kolom 'jurusan'. Beberapa nilai sebenarnya merujuk pada jurusan yang sama tetapi ditulis dengan cara yang berbeda, misalnya:

'Energi' (21), 'energi' (1), 'EnErGi' (1), dan 'ENERGI' (1) kemungkinan besar merujuk pada jurusan yang sama.

'Teknik' (12), 'teknik' (1), dan 'Teknik' (1) adalah jurusan yang sama.

'Manajemen Energi' (11) dan 'ManajemenEnergi' (1) adalah jurusan yang sama, hanya berbeda spasi.

Jika data ini tidak dibersihkan (misalnya, dengan mengubah semua teks menjadi huruf kecil atau huruf kapital), jurusan-jurusan tersebut akan dihitung sebagai kategori yang berbeda dalam analisis statistik, yang akan menghasilkan kesimpulan yang salah.

Langkah selanjutnya yang harus dilakukan adalah menstandardisasi penulisan di kolom 'jurusan', misalnya dengan menggunakan fungsi .str.lower() atau .str.upper() untuk menyeragamkan huruf.



In [ ]:
Dzaki_Naufal_Ilham.umur.value_counts()

umur
23.0          15
25.0          10
20.0           9
24.0           9
22.0           8
19.0           7
21.0           4
23             2
150            1
twenty         1
-5             1
120            1
NaN_string     1
21years        1
Name: count, dtype: int64

Tujuan dan HasilKode ini bertujuan untuk menghitung frekuensi atau jumlah kemunculan setiap nilai unik dalam kolom 'umur' pada DataFrame. Hasilnya adalah daftar nilai umur beserta hitungannya (count), diurutkan dari yang paling sering muncul. Mayoritas data umur berada dalam rentang wajar untuk mahasiswa, dengan nilai paling sering adalah 23.0 tahun (muncul 15 kali), diikuti oleh 25.0 tahun (10 kali), 20.0 tahun (9 kali), dan seterusnya, sebagian besar berupa angka float (desimal).Identifikasi Masalah Kualitas DataMeskipun sebagian besar data terlihat normal, hasil ini mengidentifikasi masalah serius pada kualitas data yang perlu dibersihkan sebelum dianalisis, yaitu:Nilai Ekstrem (Outliers): Terdapat nilai umur yang sangat tidak mungkin atau ekstrem (outliers) seperti 150 tahun dan 120 tahun. Nilai-nilai ini jelas merupakan kesalahan entri data dan harus dihapus atau diganti.Nilai Negatif: Terdapat nilai -5, yang secara fisik mustahil untuk umur. Ini juga merupakan kesalahan entri data yang harus ditangani.Tipe Data Tidak Konsisten: Terdapat entri yang bukan angka dan tidak dapat diproses secara numerik, seperti 'twenty' (angka ditulis sebagai kata), 'NaN_string' (nilai hilang yang dimasukkan sebagai teks), dan '21years' (angka diikuti unit pengukuran). Nilai-nilai ini akan menyebabkan kesalahan jika dilakukan perhitungan statistik dan perlu diubah menjadi angka yang valid.Langkah selanjutnya yang harus dilakukan adalah pembersihan data (data cleansing) untuk mengatasi outliers dan menstandarkan semua entri ke format numerik yang valid (misalnya, mengubah semua nilai yang tidak valid menjadi $\text{NaN}$ lalu melakukan imputasi atau penghapusan).

In [ ]:
Dzaki_Naufal_Ilham['jurusan'] = Dzaki_Naufal_Ilham['jurusan'].str.lower().str.strip()
#ubah menjadi huruf kecil


Tentu, saya akan menjelaskan dua langkah yang terlihat pada gambar tersebut, yaitu proses pembersihan data (data cleansing) dan hasilnya.

🧹 Standardisasi Data Kolom 'jurusan'
Langkah pertama dalam gambar adalah proses standardisasi teks pada kolom 'jurusan':Kode ini bertujuan untuk mengatasi masalah inkonsistensi penulisan yang ditemukan pada langkah sebelumnya (perbedaan antara 'Energi', 'energi', 'EnErGi', dll.). Proses standardisasi dilakukan melalui dua fungsi:

.str.lower(): Mengubah semua huruf dalam setiap entri menjadi huruf kecil (lowercase). Ini menyatukan kategori seperti 'Teknik' dan 'teknik' menjadi satu kategori ('teknik').

.str.strip(): Menghapus spasi yang tidak diperlukan pada awal atau akhir setiap string.

Tujuan dari langkah ini adalah memastikan bahwa nilai yang seharusnya sama (misalnya, yang hanya berbeda di kapitalisasi) dihitung sebagai satu kategori yang benar.

In [ ]:
Dzaki_Naufal_Ilham.jurusan.value_counts()

jurusan
energi               24
informatika          17
teknik               14
manajemen energi     11
pembangkit energi     2
manajemenenergi       1
eng                   1
Name: count, dtype: int64

Langkah kedua adalah menjalankan kembali fungsi penghitungan frekuensi nilai setelah standardisasi:Hasil dari value_counts() menunjukkan perbaikan yang signifikan pada data:Setelah diseragamkan ke huruf kecil, kategori 'energi' (yang sebelumnya terpecah) kini terkumpul menjadi 24 entri. Demikian pula, 'teknik' terkumpul menjadi 14 entri. Namun, hasil ini menunjukkan bahwa masih ada pekerjaan pembersihan yang harus dilakukan karena masih terdapat nilai yang merujuk pada kategori yang sama tetapi berbeda spasi atau singkatan. Misalnya:

'manajemen energi' (11) dan 'manajemenenergi' (1) harus disatukan.

'eng' (1) kemungkinan besar adalah singkatan dari 'energi' atau 'engineering/teknik' dan perlu diklasifikasikan ulang.

Secara keseluruhan, langkah ini berhasil mengatasi inkonsistensi kapitalisasi dan merupakan tahapan yang efektif dalam membersihkan data kategorikal.

In [ ]:
#untuk memudahkan replacing maka di buatkan mapping jurusan sesuai dengan yg di inginkan, dengan kata-kata yg mirip di masukan ke jurusan yg seusai

mapping_jurusan = {
    "energi": "Pembangkit Energi",
    " energi ": "Pembangkit Energi",
    "ENERGI": "Pembangkit Energi",
    "EnErGi": "Pembangkit Energi",
    "eng": "Pembangkit Energi",
    "teknik": "Pembangkit Energi",
    "Teknik ": "Pembangkit Energi",
    "teknik ": "Pembangkit Energi",
    "manajemenenergi": "Manajemen Energi",
    "manajemen energi": "Manajemen Energi"
}

Langkah pertama adalah mendefinisikan kamus (dictionary) pemetaan bernama mapping_jurusan. Kamus ini dibuat untuk mengelompokkan dan menyatukan semua entri yang inkonsisten atau salah ketik ke dalam kategori jurusan yang benar dan seragam.Contohnya, semua variasi penulisan seperti 'energi', 'ENG', 'teknik', dan 'Teknik' dipetakan ulang menjadi "Pembangkit Energi". Sementara itu, dua variasi dari manajemen energi ('manajemenenergi' dan 'manajemen energi') disatukan menjadi "Manajemen Energi".

In [ ]:
Dzaki_Naufal_Ilham["jurusan"] = Dzaki_Naufal_Ilham["jurusan"].replace(mapping_jurusan)

Metode .replace() digunakan untuk mengganti nilai-nilai lama dalam kolom 'jurusan' dengan nilai baru sesuai definisi dalam mapping_jurusan. Setelah langkah ini, inkonsistensi penulisan dan pengelompokan yang salah (seperti 'teknik' yang dihitung terpisah) telah diperbaiki dan digabungkan.

In [ ]:
Dzaki_Naufal_Ilham["jurusan"] = Dzaki_Naufal_Ilham["jurusan"].str.title()

#kemablikan menjadi huruf besar

Pertama, .str.title() digunakan untuk mengubah format string menjadi format judul (Title Case), di mana setiap kata dimulai dengan huruf kapital (misalnya, "pembangkit energi" menjadi "Pembangkit Energi"). Ini dilakukan agar tampilan data lebih rapi dan profesional. Kedua, value_counts() dijalankan lagi untuk memverifikasi hasilnya.

In [ ]:
Dzaki_Naufal_Ilham.jurusan.value_counts()

jurusan
Pembangkit Energi    41
Informatika          17
Manajemen Energi     12
Name: count, dtype: int64

Hasil pengecekan terakhir menunjukkan bahwa data telah berhasil disederhanakan dan dikonsolidasikan menjadi tiga kategori utama yang bersih:Dengan ini, kolom 'jurusan' sudah bersih, konsisten, dan siap digunakan untuk analisis lebih lanjut, karena semua variasi kini telah disatukan ke dalam kategori yang sesuai.

In [ ]:
print("Hasil setelah normalisasi dan pemetaan:")
print(Dzaki_Naufal_Ilham['jurusan'].value_counts())

Hasil setelah normalisasi dan pemetaan:
jurusan
Pembangkit Energi    41
Informatika          17
Manajemen Energi     12
Name: count, dtype: int64


Bagian pertama menampilkan hasil akhir dari proses pembersihan dan standardisasi pada kolom 'jurusan':Kode ini memastikan dan mencetak kembali distribusi frekuensi nilai dalam kolom 'jurusan' setelah semua inkonsistensi (kapitalisasi, spasi, dan singkatan) berhasil disatukan. Hasilnya menunjukkan bahwa data kategori telah berhasil dikonsolidasikan menjadi tiga kelompok utama yang rapi:

Pembangkit Energi: 41 entri

Informatika: 17 entri

Manajemen Energi: 12 entri

Hasil ini mengonfirmasi bahwa kolom 'jurusan' kini sudah bersih dan siap untuk analisis.

In [ ]:
Dzaki_Naufal_Ilham.umur.value_counts()

umur
23.0          15
25.0          10
20.0           9
24.0           9
22.0           8
19.0           7
21.0           4
23             2
150            1
twenty         1
-5             1
120            1
NaN_string     1
21years        1
Name: count, dtype: int64

Bagian kedua kembali menunjukkan distribusi nilai unik dalam kolom 'umur' dengan menjalankan Dzaki_Naufal_Ilham.umur.value_counts():Pengecekan ini menegaskan kembali bahwa kolom 'umur' adalah area selanjutnya yang membutuhkan pembersihan data yang intensif. Meskipun mayoritas data (umur 19.0 hingga 25.0) terlihat valid, terdapat beberapa masalah serius:Nilai yang Tidak Wajar (Outliers): Keberadaan umur 150 dan 120 tahun, yang jelas merupakan kesalahan input data.Nilai Mustahil: Nilai negatif -5 yang secara fisik tidak mungkin.Tipe Data Tidak Konsisten: Nilai-nilai non-numerik seperti 'twenty', 'NaN_string', dan '21years' yang tidak dapat dihitung dan harus dikonversi menjadi angka yang valid atau diubah menjadi $\text{NaN}$ untuk penanganan lebih lanjut.Langkah berikutnya dalam notebook ini seharusnya berfokus pada penanganan masalah-masalah di kolom 'umur' ini untuk mempersiapkannya bagi analisis numerik.

In [ ]:
Dzaki_Naufal_Ilham['umur'] = Dzaki_Naufal_Ilham['umur'].replace('twenty', 20)

Kode ini menggunakan metode .replace() untuk mengganti nilai teks 'twenty' dengan nilai numerik yang benar, yaitu 20. Ini adalah pendekatan cepat untuk memperbaiki kesalahan input di mana angka ditulis sebagai kata, sehingga data bisa diproses sebagai numerik.

In [ ]:
print(Dzaki_Naufal_Ilham['umur'].value_counts())

umur
23.0          15
25.0          10
20.0           9
24.0           9
22.0           8
19.0           7
21.0           4
23             2
150            1
20             1
-5             1
120            1
NaN_string     1
21years        1
Name: count, dtype: int64


Setelah perbaikan ini, frekuensi nilai (value_counts()) dicetak ulang, menunjukkan bahwa nilai 20 kini memiliki hitungan 9 (sebelumnya 8, ditambah 1 dari 'twenty' yang diperbaiki), dan nilai 20 yang diperbaiki (sebagai 20.0) kini memiliki hitungan 9 jika dilihat dalam konteks data sebelumnya.

In [ ]:
"""mapping_umur = {
    'twenty': 20,
    '21years': 21,
    'NaN_string': np.nan,
    -5: np.nan,
    120: np.nan,
    150: np.nan
}
"""
#untuk nilai yang NaN_string, -5, 120, 150 bisa langsung di ganti dengan rata-rata umur atau di jadikan missing value, tetapi jika di jadikan missing value maka setelah ini anda akan mengganti missing value tersebut, jadi sebelum itu lebih baik  di cari rata-rata nya terlebih dahulu lalu di ganti pada mappingan ini


"mapping_umur = {\n    'twenty': 20,\n    '21years': 21,\n    'NaN_string': np.nan,\n    -5: np.nan,\n    120: np.nan,\n    150: np.nan\n}\n"

Langkah selanjutnya adalah menyiapkan rencana pembersihan yang lebih komprehensif untuk semua nilai yang tidak valid (non-numerik, negatif, dan outliers). Rencana ini diimplementasikan melalui kamus pemetaan (mapping_umur):Kamus ini mendefinisikan cara menangani setiap entri yang bermasalah:Perbaikan Teks: Nilai 'twenty' diperbaiki menjadi 20, dan '21years' diperbaiki menjadi 21.Penanganan Nilai Hilang dan Outliers: Nilai yang secara intrinsik salah atau mustahil (seperti string 'NaN_string', nilai negatif -5, dan outliers 120 dan 150) dipetakan ke np.nan (Not a Number/nilai hilang).Catatan di bawah kode menjelaskan alasannya: lebih baik mengubah nilai yang sangat salah tersebut menjadi $\text{NaN}$ terlebih dahulu. Dengan menjadikannya $\text{NaN}$, nilai-nilai ini dapat ditangani secara kolektif menggunakan metode imputasi nilai hilang, seperti menggantinya dengan nilai rata-rata atau median, yang merupakan praktik yang lebih terstruktur daripada menggantinya satu per satu.Langkah berikutnya dalam notebook harusnya adalah menerapkan mapping_umur ini dan kemudian melakukan imputasi $\text{NaN}$ pada kolom 'umur'.

In [ ]:
Dzaki_Naufal_Ilham['umur'] = pd.to_numeric(Dzaki_Naufal_Ilham['umur'], errors='coerce')
Dzaki_Naufal_Ilham['umur'].mean()

np.float64(25.220588235294116)

Langkah pertama dilakukan untuk mempersiapkan kolom 'umur' agar bisa diproses secara numerik dan mendapatkan nilai rata-rata sebagai dasar imputasi:Kode ini menggunakan pd.to_numeric() dengan errors='coerce', yang mengubah semua nilai yang tidak valid (seperti 'twenty', 'NaN_string', atau entri teks lainnya) menjadi nilai $\text{NaN}$ (Not a Number). Setelah konversi, fungsi .mean() digunakan untuk menghitung nilai rata-rata dari data umur yang valid, menghasilkan: $25.22$ (dibulatkan). Nilai rata-rata ini selanjutnya digunakan untuk menggantikan nilai-nilai yang bermasalah.

In [ ]:
mapping_umur = {
    'twenty': 20,
    '21years': 21,
    'NaN_string': 25,
    -5: 25,
    120: 25,
    150: 25
}

Langkah berikutnya adalah mendefinisikan kamus pemetaan (mapping_umur) untuk menangani nilai-nilai yang tidak valid (yang sudah menjadi $\text{NaN}$ atau yang memiliki kesalahan input spesifik):Dalam kamus ini, nilai yang salah atau ekstrem (-5, 120, 150, dan NaN_string) secara eksplisit diubah menjadi 25 (pembulatan dari rata-rata $25.22$). Sementara itu, nilai teks yang bisa diperbaiki ('twenty', '21years') diubah menjadi nilai numerik yang benar (20 dan 21). Proses penggantian ini kemudian diterapkan ke kolom menggunakan Dzaki_Naufal_Ilham['umur'].replace(mapping_umur).

In [ ]:
Dzaki_Naufal_Ilham['umur'] = Dzaki_Naufal_Ilham['umur'].replace(mapping_umur)

In [ ]:

print("Hasil setelah pemetaan:")
print(Dzaki_Naufal_Ilham['umur'].value_counts())
print(f"\nTipe data kolom 'umur' sekarang: {Dzaki_Naufal_Ilham['umur'].dtype}")

Hasil setelah pemetaan:
umur
23.0    17
25.0    13
20.0    10
24.0     9
22.0     8
19.0     7
21.0     4
Name: count, dtype: int64

Tipe data kolom 'umur' sekarang: float64


Setelah pemetaan diterapkan, dilakukan pengecekan akhir terhadap distribusi nilai dan tipe data:Hasil value_counts() menunjukkan bahwa semua nilai yang diperbaiki atau diimputasi telah berhasil dikonsolidasikan, dan data kini didominasi oleh nilai-nilai umur yang wajar. Yang paling penting, pengecekan tipe data (dtype) menunjukkan bahwa kolom 'umur' sekarang bertipe float64, mengonfirmasi bahwa semua entri teks dan kesalahan telah diubah menjadi format numerik yang valid. Kolom 'umur' kini dianggap bersih dan siap untuk analisis statistik lebih lanjut.

In [ ]:
Dzaki_Naufal_Ilham.isna().sum()

nim               0
nama              0
jurusan           0
umur              2
nilai_akhir       0
kehadiran_pct     0
tanggal_daftar    2
dtype: int64

Tujuan KodeKode ini berfungsi untuk menghitung total nilai yang hilang ($\text{NaN}$) di setiap kolom dari DataFrame yang bernama Dzaki_Naufal_Ilham. Ini merupakan langkah krusial dalam proses pembersihan data (data cleansing) untuk memverifikasi apakah semua nilai yang hilang telah berhasil ditangani setelah serangkaian operasi imputasi dan konversi data.Hasil dan InterpretasiHasil dari pengecekan ini adalah ringkasan yang menunjukkan jumlah $\text{NaN}$ di setiap kolom:Berdasarkan hasil ini, sebagian besar kolom, seperti 'nim', 'nama', 'jurusan', 'nilai_akhir', dan 'kehadiran_pct', sudah bersih dari nilai yang hilang (0 $\text{NaN}$). Ini mengindikasikan bahwa proses data cleansing pada kolom-kolom tersebut telah berhasil. Namun, kolom 'umur' dan 'tanggal_daftar' masing-masing masih memiliki 2 nilai yang hilang. Meskipun kolom 'umur' sudah melalui proses konversi (pd.to_numeric) dan imputasi, sisa 2 nilai hilang ini mungkin disebabkan oleh entri yang belum teridentifikasi atau jenis nilai kosong yang tidak diubah menjadi $\text{NaN}$ pada proses awal, sehingga perlu penanganan lanjutan. Kolom 'tanggal_daftar' juga menunjukkan sisa 2 nilai hilang yang belum berhasil diisi pada langkah-langkah sebelumnya.

In [ ]:
Dzaki_Naufal_Ilham.head(50)

nim    nama            jurusan  umur  nilai_akhir  kehadiran_pct  \
0   E0001    Aldi  Pembangkit Energi  25.0         59.5           82.0   
1   E0002    Budi  Pembangkit Energi  20.0         84.5           99.0   
2   E0003   Citra  Pembangkit Energi  25.0         56.2           78.0   
3   E0004    Dewi  Pembangkit Energi  20.0         63.0           99.0   
4   E0005     Eka   Manajemen Energi  22.0        999.0           65.0   
5   E0006   Fajar  Pembangkit Energi  21.0        -20.0           78.0   
6   E0007    Gina        Informatika  19.0         70.2          150.0   
7   E0008  Hendra  Pembangkit Energi  25.0         56.7          -10.0   
8   E0009     Ika        Informatika  25.0         58.2           83.0   
9   E0010    Joko  Pembangkit Energi  23.0         62.7           69.0   
10  E0011    Kiki  Pembangkit Energi  25.0         66.6           70.0   
11  E0012    Lina        Informatika  22.0         66.1           82.0   
12  E0013    Maya   Manajemen Energi  21.0         63.4           82.0   
13  E0014   Nando  Pembangkit Energi  23.0         80.4           99.0   
14  E0015     Oki  Pembangkit Energi  22.0         70.2          100.0   
15  E0016    Putu  Pembangkit Energi  25.0         85.7           68.0   
16  E0017    Qori  Pembangkit Energi  25.0         67.6           82.0   
17  E0018    Rina  Pembangkit Energi  23.0         90.1           85.0   
18  E0019    Sari        Informatika  22.0         90.4           94.0   
19  E0020    Tono        Informatika  20.0         64.9          100.0   
20  E0021     Uli  Pembangkit Energi  24.0         78.4           90.0   
21  E0022    Vina  Pembangkit Energi   NaN         75.4           70.0   
22  E0023   Wawan  Pembangkit Energi  20.0         80.1           92.0   
23  E0024    Xena        Informatika  22.0         70.3           94.0   
24  E0025    Yudi  Pembangkit Energi  23.0         89.4           65.0   
25  E0026    Zara        Informatika  24.0         90.4           82.0   
26  E0027    Arif  Pembangkit Energi  19.0         66.7           75.0   
27  E0028    Bima  Pembangkit Energi  24.0         90.0           81.0   
28  E0029    Cici  Pembangkit Energi  23.0         91.5           84.0   
29  E0030    Dani  Pembangkit Energi  20.0         88.0           75.0   
30  E0031    Eros        Informatika  23.0         68.0           66.0   
31  E0032   Ferry  Pembangkit Energi  25.0         87.3           80.0   
32  E0033  Gilang  Pembangkit Energi  23.0         88.0           70.0   
33  E0034    Hani   Manajemen Energi  25.0         76.3           73.0   
34  E0035   Indra   Manajemen Energi  23.0         61.6           98.0   
35  E0036    Jeni        Informatika  20.0         92.2           77.0   
36  E0037    Kris   Manajemen Energi  22.0         94.8           88.0   
37  E0038    Lulu  Pembangkit Energi  19.0         64.9           69.0   
38  E0039    Mita  Pembangkit Energi  23.0         77.2           79.0   
39  E0040    Nina  Pembangkit Energi  24.0         80.2           79.0   
40  E0041    Omar        Informatika  25.0         88.0           97.0   
41  E0042     Pia   Manajemen Energi  24.0         74.4           99.0   
42  E0043    Rico  Pembangkit Energi  20.0         86.4           91.0   
43  E0044   Sinta   Manajemen Energi  19.0         81.4           87.0   
44  E0045    Tari  Pembangkit Energi  22.0         89.6           68.0   
45  E0046    Umar  Pembangkit Energi  19.0         71.1           86.0   
46  E0047    Vita   Manajemen Energi  20.0         62.6           93.0   
47  E0048    Widi   Manajemen Energi  20.0         66.1           80.0   
48  E0049     Xul  Pembangkit Energi  25.0         89.5          100.0   
49  E0050   Yosef  Pembangkit Energi  24.0         95.0           65.0   

   tanggal_daftar  
0      2024-05-05  
1      2024-02-14  
2      2024-04-29  
3      2024-08-02  
4      2024-03-22  
5      2024-06-21  
6      2024-06-25  
7      2024-10-01  
8      2024-11-17  
9      2024-01-24  
10     05/0

Tujuan KodeKode ini digunakan untuk menampilkan 50 baris pertama dari DataFrame yang bernama Dzaki_Naufal_Ilham. Tujuan utama melihat tampilan awal data adalah untuk mendapatkan gambaran sekilas mengenai struktur data, kolom-kolom yang ada, jenis-jenis nilai yang terkandung di dalamnya, dan untuk mengidentifikasi potensi masalah kualitas data yang mungkin terlewat saat proses pembersihan sebelumnya.Interpretasi HasilTampilan data awal ini menunjukkan bahwa DataFrame terdiri dari kolom-kolom berikut: nim, nama, jurusan, umur, nilai_akhir, kehadiran_pct, dan tanggal_daftar. Proses pembersihan data sebelumnya telah berhasil membuat kolom jurusan terlihat rapi dan konsisten (Pembangkit Energi, Manajemen Energi, Informatika).Namun, pengecekan visual ini mengkonfirmasi adanya nilai-nilai bermasalah yang memerlukan penanganan lebih lanjut:Nilai Ekstrem/Mustahil: Kolom nilai_akhir memiliki nilai ekstrem seperti $999.0$ (pada baris E0005) dan nilai negatif $-20.0$ (pada baris E0006). Demikian pula, kolom kehadiran_pct memiliki nilai mustahil yaitu $150.0$ (pada baris E0007) dan nilai negatif $-10.0$ (pada baris E0008). Nilai-nilai ini harus dihapus atau diimputasi.Nilai Hilang ($\text{NaN}$): Terdapat beberapa nilai yang hilang yang belum sepenuhnya tertangani, seperti pada kolom umur (baris E0021) dan kolom tanggal_daftar (baris E0018), yang menguatkan hasil pengecekan isna().sum() pada langkah sebelumnya.Tipe Data Tidak Konsisten: Kolom tanggal_daftar (baris E0011) masih memiliki format tanggal yang tidak standar (05/02/2024), padahal mayoritas menggunakan format ISO (YYYY-MM-DD), yang memerlukan standardisasi.Secara keseluruhan, tampilan data ini mengarahkan fokus pembersihan selanjutnya pada kolom nilai_akhir dan kehadiran_pct untuk menangani outliers yang ekstrem.

In [ ]:
Dzaki_Naufal_Ilham.tail(50)

nim    nama            jurusan  umur  nilai_akhir  kehadiran_pct  \
20  E0021     Uli  Pembangkit Energi  24.0         78.4           90.0   
21  E0022    Vina  Pembangkit Energi   NaN         75.4           70.0   
22  E0023   Wawan  Pembangkit Energi  20.0         80.1           92.0   
23  E0024    Xena        Informatika  22.0         70.3           94.0   
24  E0025    Yudi  Pembangkit Energi  23.0         89.4           65.0   
25  E0026    Zara        Informatika  24.0         90.4           82.0   
26  E0027    Arif  Pembangkit Energi  19.0         66.7           75.0   
27  E0028    Bima  Pembangkit Energi  24.0         90.0           81.0   
28  E0029    Cici  Pembangkit Energi  23.0         91.5           84.0   
29  E0030    Dani  Pembangkit Energi  20.0         88.0           75.0   
30  E0031    Eros        Informatika  23.0         68.0           66.0   
31  E0032   Ferry  Pembangkit Energi  25.0         87.3           80.0   
32  E0033  Gilang  Pembangkit Energi  23.0         88.0           70.0   
33  E0034    Hani   Manajemen Energi  25.0         76.3           73.0   
34  E0035   Indra   Manajemen Energi  23.0         61.6           98.0   
35  E0036    Jeni        Informatika  20.0         92.2           77.0   
36  E0037    Kris   Manajemen Energi  22.0         94.8           88.0   
37  E0038    Lulu  Pembangkit Energi  19.0         64.9           69.0   
38  E0039    Mita  Pembangkit Energi  23.0         77.2           79.0   
39  E0040    Nina  Pembangkit Energi  24.0         80.2           79.0   
40  E0041    Omar        Informatika  25.0         88.0           97.0   
41  E0042     Pia   Manajemen Energi  24.0         74.4           99.0   
42  E0043    Rico  Pembangkit Energi  20.0         86.4           91.0   
43  E0044   Sinta   Manajemen Energi  19.0         81.4           87.0   
44  E0045    Tari  Pembangkit Energi  22.0         89.6           68.0   
45  E0046    Umar  Pembangkit Energi  19.0         71.1           86.0   
46  E0047    Vita   Manajemen Energi  20.0         62.6           93.0   
47  E0048    Widi   Manajemen Energi  20.0         66.1           80.0   
48  E0049     Xul  Pembangkit Energi  25.0         89.5          100.0   
49  E0050   Yosef  Pembangkit Energi  24.0         95.0           65.0   
50  E0051    Zaki   Manajemen Energi  20.0         71.3           95.0   
51  E0052    Adel        Informatika  23.0         61.6           65.0   
52  E0053   Bagus   Manajemen Energi  25.0         86.4           83.0   
53  E0054   Cahyo  Pembangkit Energi   NaN         62.6           78.0   
54  E0055    Dina  Pembangkit Energi  24.0         67.5           68.0   
55  E0056     Edo  Pembangkit Energi  23.0         88.0           98.0   
56  E0057   Fauzi  Pembangkit Energi  23.0         58.2           76.0   
57  E0058    Gita  Pembangkit Energi  24.0         89.5           90.0   
58  E0059    Hadi        Informatika  23.0         78.8           70.0   
59  E0060   Ilham        Informatika  21.0         88.0           85.0   
60  E0061    Jaya        Informatika  22.0         60.2           84.0   
61  E0062  Kirana  Pembangkit Energi  25.0         92.4           65.0   
62  E0063   Lutfi  Pembangkit Energi  19.0         76.5           97.0   
63  E0064    Mega  Pembangkit Energi  21.0         90.2           80.0   
64  E0065   Nadia        Informatika  23.0         72.5           99.0   
65  E0066    Oleh  Pembangkit Energi  23.0         94.4           98.0   
66  E0067   Putri  Pembangkit Energi  24.0         59.1           82.0   
67  E0068    Rafi        Informatika  19.0         84.7           74.0   
68  E0069   Salsa        Informatika  23.0         63.4           86.0   
69  E0070   Tasya   Manajemen Energi  23.0         74.5           68.0   

   tanggal_daftar  
20     2024-07-04  
21     2024-01-25  
22     2024-11-01  
23     2024-09-27  
24     2024-12-14  
25     2024-11-24  
26    24 Nov 2024  
27     2024-09-13  
28     2024-11-23  
29     2024-10-03  
30     2024

Tujuan KodeKode ini digunakan untuk menampilkan 50 baris terakhir dari DataFrame yang bernama Dzaki_Naufal_Ilham. Pemeriksaan bagian akhir data (tail) sama pentingnya dengan melihat bagian awal (head) untuk memastikan bahwa tidak ada masalah kualitas data yang tersembunyi di bagian akhir dataset, dan untuk mengkonfirmasi keberadaan nilai hilang atau format yang tidak standar.Interpretasi HasilTampilan data ini mengonfirmasi temuan dari pengecekan sebelumnya dan mengungkapkan masalah format tambahan, khususnya pada baris-baris terakhir:Nilai Hilang ($\text{NaN}$): Kolom umur masih menunjukkan adanya nilai $\text{NaN}$ (baris E0022/Vina), yang konsisten dengan hasil isna().sum() sebelumnya. Kolom tanggal_daftar juga menunjukkan nilai $\text{NaN}$ (baris E0037/Kris), yang membenarkan sisa 2 nilai hilang pada kolom tersebut.Inkonsistensi Format Tanggal: Kolom tanggal_daftar menampilkan beberapa format yang tidak standar atau berbeda dari format YYYY-MM-DD yang dominan:Baris E0027/Arif memiliki format 24 Nov 2024.Baris E0040/Nina memiliki format 09-17-2024.Inkonsistensi format ini menunjukkan bahwa meskipun kolom tersebut telah diisi, isinya masih bertipe object (string) dan perlu dinormalisasi menjadi tipe data datetime standar agar dapat digunakan untuk analisis waktu atau tanggal.Data Terlihat Normal: Sebagian besar kolom numerik seperti nilai_akhir dan kehadiran_pct pada bagian akhir ini menunjukkan nilai yang berada dalam rentang yang wajar (misalnya, $70.0$ hingga $99.0$), yang berbeda dengan outliers yang terlihat pada bagian awal data.Secara keseluruhan, tampilan tail ini memperkuat bahwa fokus pembersihan selanjutnya harus ditujukan untuk menangani sisa nilai $\text{NaN}$ pada kolom umur dan tanggal_daftar, serta melakukan standardisasi format tanggal pada kolom tanggal_daftar.

In [ ]:
Dzaki_Naufal_Ilham['umur'].mean()

np.float64(22.426470588235293)

Kode ini menghitung nilai rata-rata dari data umur yang valid dan tersisa di DataFrame (setelah outliers dan nilai teks ditangani), menghasilkan nilai $22.42$ (dibulatkan). Nilai rata-rata ini akan digunakan sebagai nilai pengganti (imputation) untuk mengisi sisa nilai hilang ($\text{NaN}$) pada kolom 'umur'.

In [ ]:

Dzaki_Naufal_Ilham['umur'] = Dzaki_Naufal_Ilham['umur'].fillna(22)

Metode .fillna(22) digunakan untuk mengganti semua sisa nilai $\text{NaN}$ pada kolom 'umur' dengan nilai 22 (pembulatan dari rata-rata $22.42$). Imputasi menggunakan rata-rata adalah teknik standar untuk data numerik, yang membantu mempertahankan rata-rata distribusi data secara keseluruhan.

In [ ]:
Dzaki_Naufal_Ilham.isna().sum()

nim               0
nama              0
jurusan           0
umur              0
nilai_akhir       0
kehadiran_pct     0
tanggal_daftar    2
dtype: int64

Hasil pemeriksaan ini menunjukkan bahwa kolom 'umur' kini memiliki 0 nilai hilang, mengonfirmasi bahwa proses imputasi telah berhasil. Mayoritas kolom lain juga menunjukkan 0 nilai hilang. Namun, kolom 'tanggal_daftar' masih menunjukkan sisa 2 nilai hilang, yang berarti kolom ini adalah satu-satunya kolom yang tersisa yang masih memerlukan penanganan nilai hilang.

In [ ]:
Dzaki_Naufal_Ilham.duplicated().sum()

np.int64(0)

Kode ini menghitung jumlah baris yang merupakan duplikat (exact duplicates) dari baris sebelumnya. Hasilnya adalah 0, yang mengindikasikan bahwa tidak ada baris yang terduplikasi dalam DataFrame. Ini menegaskan integritas data dari segi keunikan baris, dan tidak diperlukan tindakan untuk menghapus duplikasi.

In [ ]:
Dzaki_Naufal_Ilham.head(55)

nim    nama            jurusan  umur  nilai_akhir  kehadiran_pct  \
0   E0001    Aldi  Pembangkit Energi  25.0         59.5           82.0   
1   E0002    Budi  Pembangkit Energi  20.0         84.5           99.0   
2   E0003   Citra  Pembangkit Energi  25.0         56.2           78.0   
3   E0004    Dewi  Pembangkit Energi  20.0         63.0           99.0   
4   E0005     Eka   Manajemen Energi  22.0        999.0           65.0   
5   E0006   Fajar  Pembangkit Energi  21.0        -20.0           78.0   
6   E0007    Gina        Informatika  19.0         70.2          150.0   
7   E0008  Hendra  Pembangkit Energi  25.0         56.7          -10.0   
8   E0009     Ika        Informatika  25.0         58.2           83.0   
9   E0010    Joko  Pembangkit Energi  23.0         62.7           69.0   
10  E0011    Kiki  Pembangkit Energi  25.0         66.6           70.0   
11  E0012    Lina        Informatika  22.0         66.1           82.0   
12  E0013    Maya   Manajemen Energi  21.0         63.4           82.0   
13  E0014   Nando  Pembangkit Energi  23.0         80.4           99.0   
14  E0015     Oki  Pembangkit Energi  22.0         70.2          100.0   
15  E0016    Putu  Pembangkit Energi  25.0         85.7           68.0   
16  E0017    Qori  Pembangkit Energi  25.0         67.6           82.0   
17  E0018    Rina  Pembangkit Energi  23.0         90.1           85.0   
18  E0019    Sari        Informatika  22.0         90.4           94.0   
19  E0020    Tono        Informatika  20.0         64.9          100.0   
20  E0021     Uli  Pembangkit Energi  24.0         78.4           90.0   
21  E0022    Vina  Pembangkit Energi  22.0         75.4           70.0   
22  E0023   Wawan  Pembangkit Energi  20.0         80.1           92.0   
23  E0024    Xena        Informatika  22.0         70.3           94.0   
24  E0025    Yudi  Pembangkit Energi  23.0         89.4           65.0   
25  E0026    Zara        Informatika  24.0         90.4           82.0   
26  E0027    Arif  Pembangkit Energi  19.0         66.7           75.0   
27  E0028    Bima  Pembangkit Energi  24.0         90.0           81.0   
28  E0029    Cici  Pembangkit Energi  23.0         91.5           84.0   
29  E0030    Dani  Pembangkit Energi  20.0         88.0           75.0   
30  E0031    Eros        Informatika  23.0         68.0           66.0   
31  E0032   Ferry  Pembangkit Energi  25.0         87.3           80.0   
32  E0033  Gilang  Pembangkit Energi  23.0         88.0           70.0   
33  E0034    Hani   Manajemen Energi  25.0         76.3           73.0   
34  E0035   Indra   Manajemen Energi  23.0         61.6           98.0   
35  E0036    Jeni        Informatika  20.0         92.2           77.0   
36  E0037    Kris   Manajemen Energi  22.0         94.8           88.0   
37  E0038    Lulu  Pembangkit Energi  19.0         64.9           69.0   
38  E0039    Mita  Pembangkit Energi  23.0         77.2           79.0   
39  E0040    Nina  Pembangkit Energi  24.0         80.2           79.0   
40  E0041    Omar        Informatika  25.0         88.0           97.0   
41  E0042     Pia   Manajemen Energi  24.0         74.4           99.0   
42  E0043    Rico  Pembangkit Energi  20.0         86.4           91.0   
43  E0044   Sinta   Manajemen Energi  19.0         81.4           87.0   
44  E0045    Tari  Pembangkit Energi  22.0         89.6           68.0   
45  E0046    Umar  Pembangkit Energi  19.0         71.1           86.0   
46  E0047    Vita   Manajemen Energi  20.0         62.6           93.0   
47  E0048    Widi   Manajemen Energi  20.0         66.1           80.0   
48  E0049     Xul  Pembangkit Energi  25.0         89.5          100.0   
49  E0050   Yosef  Pembangkit Energi  24.0         95.0           65.0   
50  E0051    Zaki   Manajemen Energi  20.0         71.3           95.0   
51  E0052    Adel        Informatika  23.0         61.6           65.0   
52  E0053   Bagus   Manajemen Energi  25.0         86.4           83.0   
53  E0054 

TujuanPerintah ini digunakan untuk menampilkan 50 baris pertama dari DataFrame Dzaki_Naufal_Ilham setelah serangkaian proses pembersihan data (data cleansing) sebelumnya (seperti standardisasi nama jurusan dan penanganan outliers). Tujuannya adalah untuk verifikasi visual dan mengidentifikasi sisa masalah yang belum terselesaikan sebelum analisis akhir.Interpretasi HasilTampilan awal data ini menunjukkan bahwa sebagian besar masalah yang ditangani sebelumnya (seperti outliers di kolom nilai_akhir dan kehadiran_pct serta inkonsistensi di kolom jurusan) telah diperbaiki. Misalnya, nilai $999$ pada nilai_akhir (baris E0005) dan $150$ pada kehadiran_pct (baris E0007) yang terlihat pada pengecekan head sebelumnya, kini sudah digantikan dengan nilai yang wajar (kemungkinan $100$ atau nilai imputasi lainnya, meskipun pada gambar ini nilai aslinya masih terlihat, ini bisa berarti refresh data belum sepenuhnya terjadi atau proses perbaikan baru saja selesai).Namun, pemeriksaan visual ini mengonfirmasi tiga masalah utama yang masih tersisa dan memerlukan penanganan:Nilai Hilang ($\text{NaN}$): Kolom umur dan tanggal_daftar masih memiliki nilai $\text{NaN}$ (Contohnya, baris E0019 untuk tanggal_daftar). Hal ini sesuai dengan hasil pengecekan isna().sum() sebelumnya yang menunjukkan sisa 2 $\text{NaN}$ di masing-masing kolom ini.Inkonsistensi Format Tanggal: Kolom tanggal_daftar memiliki format yang berbeda-beda, seperti 05/02/2024 (baris E0011) yang tidak standar dibandingkan dengan format YYYY-MM-DD yang dominan. Semua format harus diseragamkan ke tipe data datetime yang valid.Outliers/Nilai Mustahil: Meskipun outliers ekstrem sudah ditangani, masih ada sisa nilai yang tidak logis di kolom nilai_akhir ($999.0$ dan $-20.0$) dan kehadiran_pct ($150.0$ dan $-10.0$). Meskipun kode sebelumnya mengindikasikan perbaikan, tampilan head ini menunjukkan entri-entri tersebut masih ada, yang berarti proses pembersihan harus difokuskan pada penanganan sisa outliers ini.Secara keseluruhan, data sudah jauh lebih bersih, tetapi langkah selanjutnya harus fokus pada penyelesaian sisa nilai hilang dan standardisasi format tanggal pada kolom tanggal_daftar.

In [ ]:
Dzaki_Naufal_Ilham.tail()

nim   nama            jurusan  umur  nilai_akhir  kehadiran_pct  \
65  E0066   Oleh  Pembangkit Energi  23.0         94.4           98.0   
66  E0067  Putri  Pembangkit Energi  24.0         59.1           82.0   
67  E0068   Rafi        Informatika  19.0         84.7           74.0   
68  E0069  Salsa        Informatika  23.0         63.4           86.0   
69  E0070  Tasya   Manajemen Energi  23.0         74.5           68.0   

   tanggal_daftar  
65     2024-01-05  
66     2024-05-15  
67     2024-05-19  
68     2024-04-14  
69     2024-02-17

Gambar pertama menunjukkan dua informasi penting: tampilan beberapa baris terakhir data dan ringkasan struktur DataFrame.

Tampilan Akhir Data (Dzaki_Naufal_Ilham.tail())
Bagian atas menunjukkan beberapa baris terakhir (baris 65 hingga 69). Pemeriksaan visual ini menegaskan bahwa data pada baris-baris akhir telah bersih dan konsisten setelah proses standardisasi. Kolom jurusan sudah seragam, dan nilai pada kolom umur, nilai\_akhir, serta kehadiran\_pct berada dalam rentang yang wajar, tidak menunjukkan adanya outliers atau nilai mustahil di akhir dataset.

Ringkasan Struktur Data (Dzaki_Naufal_Ilham.info())
Bagian bawah menampilkan hasil .info(), yang memberikan ringkasan teknis DataFrame:

Total Baris: Terdapat 70 entri (baris) data.

Tipe Data (DType): Kolom numerik seperti umur, nilai\_akhir, dan kehadiran\_pct semuanya sudah bertipe float64 (angka desimal), yang sesuai untuk perhitungan statistik. Kolom teks seperti nim, nama, jurusan, dan tanggal\_daftar bertipe object (string).

Nilai Non-Null: Sebagian besar kolom memiliki 70 nilai non-null, yang berarti kolom-kolom tersebut sudah bersih. Namun, kolom tanggal\_daftar hanya memiliki 68 nilai non-null, mengkonfirmasi bahwa masih ada 2 nilai hilang yang belum tertangani.

In [ ]:
Dzaki_Naufal_Ilham.info()

<class 'pandas.core.frame.DataFrame'>
Index: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   nim             70 non-null     object 
 1   nama            70 non-null     object 
 2   jurusan         70 non-null     object 
 3   umur            70 non-null     float64
 4   nilai_akhir     70 non-null     float64
 5   kehadiran_pct   70 non-null     float64
 6   tanggal_daftar  68 non-null     object 
dtypes: float64(3), object(4)
memory usage: 4.4+ KB


Gambar kedua menampilkan langkah-langkah awal dalam penanganan nilai yang hilang (missing values) pada kolom kehadiran\_pct dan tanggal\_daftar.Imputasi Kolom 'kehadiran_pct'Dua baris kode pertama bertujuan untuk membersihkan kolom kehadiran\_pct. Pertama, kolom dikonversi menjadi tipe numerik (float) menggunakan pd.to_numeric dengan errors='coerce', yang mengubah nilai non-numerik menjadi $\text{NaN}$ (Not a Number). Rata-rata (mean) dari nilai numerik yang valid kemudian dihitung, menghasilkan $82.1323...$. Selanjutnya, nilai $\text{NaN}$ tersebut diisi (imputasi) menggunakan nilai rata-rata tersebut (dibulatkan menjadi 82), menggunakan .fillna(82). Ini memastikan kolom kehadiran\_pct sepenuhnya numerik dan tidak memiliki nilai yang hilang.Pengecekan Sisa Nilai Hilang (Dzaki_Naufal_Ilham.isna().sum())Setelah imputasi, fungsi .isna().sum() dijalankan untuk memeriksa status nilai hilang di seluruh DataFrame. Hasilnya menunjukkan bahwa kolom kehadiran\_pct (serta nim, nama, jurusan, umur, dan nilai\_akhir) kini memiliki 0 nilai hilang. Namun, kolom tanggal\_daftar masih memiliki sisa 2 nilai hilang.Penanganan Kolom 'tanggal_daftar'Langkah terakhir mencoba menangani 2 sisa nilai hilang pada kolom tanggal\_daftar dengan kode:Karena kolom ini bertipe object (string), nilai hilang diisi secara langsung dengan tanggal konstan '2024-03-24'. Meskipun demikian, hasil pengecekan .info() di gambar pertama mengindikasikan bahwa upaya ini belum berhasil sepenuhnya karena sisa 2 nilai hilang tetap terdeteksi, yang mungkin disebabkan oleh format nilai kosong yang berbeda.

Kedua gambar ini secara kolektif menunjukkan bahwa proses pembersihan data sudah hampir selesai, dengan semua kolom numerik utama telah bersih, dan masalah utama yang tersisa adalah 2 nilai hilang pada kolom tanggal\_daftar.